### PR TITLE
journal: Add TrustedFieldAllowList option

### PR DESCRIPTION
**Pull request by daandemeyer**

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
Hmmm, I'm testing this out and getting some surprising results. If I only log the _BOOT_ID trusted field into the journal, the number of fields logged is reduced as expected but I get double the number of DATA and ENTRY objects in the journal. I'd expect the number of DATA objects to reduce as well and the number of ENTRY objects to stay the same. Anyone have an idea on which trusted field could be causing this increase by omitting it from the journal?

With TrustedFieldAllowList set to "_BOOT_ID":

```
OBJECT TYPE         ENTRIES   SIZE   
Unused                  0                0B     
Data                       4822940   366.9M 
Field                       187            9.2K   
Entry                      4822405   662.2M 
Data Hash Table  28              99.5M  
Field Hash Table  28              146.1K 
Entry Array           1907          273.3M 
Tag                         0                0B     
```

Without the option set:

```
OBJECT TYPE         ENTRIES  SIZE
Unused                  0               0B
Data                       2197334  166.6M
Field                       316           15.5K
Entry                      2196147   804.3M
Data Hash Table  13              46.2M
Field Hash Table  13              67.8K
Entry Array           3468          507.7M
Tag                         0                0B
```

File: src/basic/io-util.h [hunk 1]
```diff
@@ -79,6 +82,15 @@ static inline bool FILE_SIZE_VALID_OR_INFINITY(uint64_t l) {
 #define IOVEC_INIT_STRING(string) IOVEC_INIT((char*) string, strlen(string))
 #define IOVEC_MAKE_STRING(string) (struct iovec) IOVEC_INIT_STRING(string)
 
+static inline struct iovec *iovec_free(struct iovec *iovec) {
+        free(iovec->iov_base);
```
Comment (vcaputo) [inline src/basic/io-util.h:89]: This needs to be conditional on iovec being non-NULL

**Comment by vcaputo** [comment]
Just took a casual look, not a fan of either the approach taken or the feature altogether really.

**Comment by DaanDeMeyer** [comment]
Turns out I just didn't think my test through. My test service was logging as much as it could for 120s. When logging fewer fields journald also starts processing log messages a lot faster which led to a lot more messages being logged overall, explaining the size increase. Logging a fixed amount of data gives the expected results. If I only log the _BOOT_ID trusted field the journal's size reduces to less then half its size compared to when we log all trusted fields.

**Comment by DaanDeMeyer** [comment]
I'm going to switch to the original idea where we don't add fields we want to filter to the iovec in the first place. That allows us to filter all fields added by journald without touching fields added by services (which could potentially have the same name as journald's fields. It's also more efficient as we're not working with a second array.

**Comment by anitazha** [comment]
> Just took a casual look, not a fan of neither the approach taken or the feature altogether really.

@vcaputo Could I ask why? I've seen that you've done some work on the journal so I'm wondering if you have some concerns around this. We're looking to optimize the journal size as much as possible and from Daan's tests this is kind of a low hanging fruit in terms of efficiency in processing and space.

**Comment by vcaputo** [comment]
> > Just took a casual look, not a fan of neither the approach taken or the feature altogether really.
> 
> @vcaputo Could I ask why? I've seen that you've done some work on the journal so I'm wondering if you have some concerns around this. We're looking to optimize the journal size as much as possible and from Daan's tests this is kind of a low hanging fruit in terms of efficiency in processing and space.

WRT implementation: just don't acquire the filtered out fields in the first place, as mentioned by @poettering.

WRT feature altogether: this strikes me as a foot-gun for people to break both existing functionality and existing documentation on how to work with journals.  *Everything* up until now has been able to expect these fields would be present.

If the objective is to minimize space consumption, I don't think simply throwing away information is the appropriate approach at all.

In terms of space-saving low-hanging fruit, how about debugging and landing the reverted truncating of archived journals?  After that's sorted out, add punching holes in the unused parts of entry arrays.  They grow exponentially; the last entry array in every chain is both the largest and has a good chance of being half or mostly empty.

To go further down the space saving route I think some serious thought needs to be given to compressing more of the journal which tends to be highly compressible.  Just taking my server's 96M journal here and running it through plain gzip produces a 14M file, *losslessly*.

The way I view the journal's current state is it has established a base of functionality and interfaces that have proven relatively stable and desirable, which the world now depend upon.  Now it's time to improve the quality of how it stores and accesses the information, iterating in that backend area on both space and speed axes.  *NOT* by reducing how much information is being stored, that's a lazy cop-out approach which threatens to disrupt arguably one of the few redeeming qualities in the current state of affairs.

File: src/basic/hash-funcs.c [hunk 2]
```diff
@@ -16,6 +17,33 @@ DEFINE_HASH_OPS_FULL(string_hash_ops_free_free,
...
 
+void iovec_hash_func(const struct iovec *iovec, struct siphash *state) {
+        siphash24_compress(iovec->iov_base, iovec->iov_len, state);
+}
+
+int iovec_compare_func(const struct iovec *a, const struct iovec *b) {
+        assert(a);
+        assert(b);
+
+        int c = memcmp(a->iov_base, b->iov_base, MIN(a->iov_len, b->iov_len));
```
Comment (poettering) [inline src/basic/hash-funcs.c:28]: memcmp_nn() is your friend

File: src/basic/io-util.c [hunk 3]
```diff
@@ -277,6 +278,37 @@ char* set_iovec_string_field_free(struct iovec *iovec, size_t *n_iovec, const ch
...
+        assert(s);
+
+        iovec = new(struct iovec, 1);
+        if (!iovec)
+                return -ENOMEM;
+
+        size_t len = strlen(s);
+
+        *iovec = (struct iovec) {
+                .iov_base = memdup(s, len),
```
Comment (poettering) [inline src/basic/io-util.c:296]: hmm, is it really worth dropping the suffix \0?

**Comment by poettering** [comment]
i don't like this approach: we should filter after acquiring the data, but not even acquire the data. Acquiring the data is slow after all (procfs accesses, …) hence if we add this we should add it in a way that people who are concerned about speed (and not disk space) can benefit too.

please add an enum for every field we add, plus a string table for it. Then define a bitmask type with one bit for each of these enums, and a parser to convert a string (supplied via getenv) into such a bitmask. Then, when acquiring the metadta alaways check the bitmask before we acquire the metadata whether we actually should do so.

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by DaanDeMeyer** [comment]
I'm going to close this in favor of adding an option that allows configuring which fields should be indexed instead. @vcaputo is right that allowing to remove arbitrary fields entirely might cause issues. Allowing configurable indexing information should be a lot safer I think and should still result in some good savings given a third of the journal's size on my laptop consists of entry arrays.

```
daandemeyer@ddm ~/p/systemd (analyze-journal)> build/systemd-analyze journal
OBJECT TYPE      ENTRIES SIZE  
Unused           0       0B
Data             3655870 646.3M
Field            3692    195.2K
Entry            4050199 1.5G
Data Hash Table  32      113.7M
Field Hash Table 32      167.0K
Entry Array      484732  1.1G
Tag              0       0B
daandemeyer@ddm ~/p/systemd (analyze-journal)> journalctl --system --disk-usage
Archived and active journals take up 3.4G in the file system.
```

**Comment by vcaputo** [comment]
@DaanDeMeyer FYI I've added some more elaborate EntryArrayObject accounting than systemd-analyze provides to [jio](https://github.com/vcaputo/jio/blob/master/src/report-entry-arrays.c), which might be useful in quantifying how much space would be reclaimed by punching holes in the allocated-but-unused portions of the journal entry arrays when archiving.

It also identifies how many entry arrays are identical within a journal, which could be reclaimed if the file format were changed to support sharing entry array chains in some fashion until they deviate... which might be interesting from not just a space-saving perspective but also cache utilization during filtering of entries by multiple data objects.  This aspect is less clear of a potential win, especially considering the complexity of such a format change.

The hole-punching however I think is a no-brainer, given the exponential growth pattern... on my system the entry arrays utilization is often around 50% += 10%.  So your 1.1G example on my system would reclaim around half of that on rotate if the unused portions of the entry arrays were perforated to give the space back to the FS.

**Pull request closed without merging**